spec/index-spec.ts

```diff
@@ -116,93 +116,87 @@ describe('array', () => {
 describe('object', () => {
   describe('when given valid JSON', () => {
     it('can decode a simple object', () => {
-      const decoder = module.object(
-        ['x', module.number()],
-        (x) => ({x})
-      )
+      const decoder = module.object({
+        x: module.field('x', module.number())
+      })
       expect(decoder.decodeJSON('{"x": 5}')).toEqual({x: 5})
     })
 
     it('can decode a nested object', () => {
-      const decoder = module.object(
-        ['payload', module.object(
-          ['x', module.number()],
-          ['y', module.number()],
-          (x, y) => ({x, y})
-        )],
-        ['error', module.equal(false)],
-        (payload, error) => ({payload, error})
-      )
+      const decoder = module.object({
+        payload: module.field(
+          'payload',
+          module.object({
+            x: module.field('x', module.number()),
+            y: module.field('y', module.number())
+          })
+        ),
+        error: module.field('error', module.equal(false))
+      })
       const json = '{"payload": {"x": 5, "y": 2}, "error": false}'
       expect(decoder.decodeJSON(json)).toEqual({
         payload: {x: 5, y: 2},
         error: false
       })
     })
 
-    it('can decode into something other than an object', () => {
-      const decoder = module.object(
-        ['x', module.number()],
-        ['y', module.number()],
-        (x, y) => [x, y]
-      )
-      expect(decoder.decodeJSON('{"x":3,"y":4}')).toEqual([3, 4])
+    it('can reach into nested objects', () => {
+      const decoder = module.object({
+        token: module.field('token', module.string()),
+        name: module.at(['person', 'name'], module.string()),
+        age: module.at(['person', 'age'], module.number())
+      })
+      const json = '{"token":"abcdef","person":{"name":"John","age":21}}'
+      expect(decoder.decodeJSON(json)).toEqual({
+        token: 'abcdef',
+        name: 'John',
+        age: 21
+      })
     })
   })
 
   describe('when incorrect JSON', () => {
     it('fails when not given an object', () => {
-      const decoder = module.object(
-        ['x', module.number()],
-        (x: number) => ({x})
-      )
+      const decoder = module.object({
+        x: module.field('x', module.number())
+      })
       expect(() => decoder.decodeJSON('true')).toThrowError(
         `error at root: expected object, got boolean`
       )
     })
 
     it('reports a missing key', () => {
-      const decoder = module.object(
-        ['x', module.number()],
-        (x: number) => ({x})
-      )
+      const decoder = module.object({
+        x: module.field('x', module.number())
+      })
       expect(() => decoder.decodeJSON('{}')).toThrowError(
-        `error at root: expected object with keys: x`
-      )
-    })
-
-    it('reports multiple missing keys in alphabetical order', () => {
```

Comment:
I really should have written a little bit more about the changes I made when I submitted this pull request, but it was friday afternoon and I just wanted to get it out there...
Anyway, the reason that this test went away is that using the `field` decoder in the object definition, it is not really possible anymore to do this because missing fields are not checked in the `object`, but in the `field` decoder. I choose this implementation because I like they way Elm implemented this. You can write more flexible object decoders that reach into the JSON structure, e.g.

```typescript
object({
    name: field('name', string()),
    age: at(['info', 'age'], number()),
    height: at(['info', 'height'], number()),
})
```

But I understand that checking all the fields first would be nice, too, so in retrospect, I saw that we can probably combine the new mapped types approach with the previous object definition using tuples, that would look like this:

```typescript
object({
    x: ['x', number()],
    y: ['y', number()],
})
```

This wouldn't be as flexible, but we could add back in the missing keys check for the whole object.